### DISCUSSION TITLE
How would you get enity under the cursor when mouse clicked?

### DISCUSSION THREAD

**Discussion by pwmit** [AlmasB/FXGL#1294] [Q&A]
When I click the mouse over an entiy is there a way to return whats entity is under the cursor rather than adding events to the entity builder? Also are you able to get the with(mycomponent) instance from the entity?
I've been using code below but doesn't seem very efficient. It's creating a handler every instance and there could be several hundred entity in the game world.

`   entity.getViewComponent().addOnClickHandler(event -> {
                    buildingClick(event, entity);
                });
				`

I have noticed theres a IsCollision method which could be used? Just for context the game is an RTS and when you click on people, troops, tanks etc I need to be able to give them directions. Also when I click on a building I need to see it's options and info.

**Comment by AlmasB** [comment]
Hi, one possible way would be if entities have hit boxes, then `gameWorld.getEntitiesInRange()` using `input.getMousePositionWorld()` as a basis. The first method takes a rectangle, so you can use the mouse position and a suitable width/height combination to select units.

As for component instance, yes:

```
entity.getComponent(MyComponent.class) gives you the instance of MyComponent
```

Let me know if any of this works.

**Reply by pwmit** [reply]
Yes this works well thank you.


`Class Road extends BuildingParent {}

........

Cursor class...

void mouseClick(MouseEvent mouseEvent) {
  Rectangle2D r = new Rectangle2D(mousePos.getX()-4,mousePos.getY()-4,mousePos.getX()+4,mousePos.getY()+4);
  List el = FXGL.getGameWorld().getEntitiesInRange(r);
  for(Entity e : el) {
    if(e.hasComponent(Road.class)) {
    Log.i("road " +e.getComponent(Road.class));
    }
    else if(e.hasComponent(CarPark.class)) {
    Log.i("road " +e.getComponent(CarPark.class));
  }
}

}`

The only problem here is that I would have to check for each component individually. I'm not sure there would be a way to get it via the parent class  BuildingParent.class

**Reply by pwmit** [reply]
correction that should be  new Rectangle2D(mousePos.getX()-4,mousePos.getY()-4,8,8);
